### PR TITLE
blockstream-electrs: init at 0.4.1-unstable-2024-11-25

### PR DESCRIPTION
**Pull request by phlip9**
## Description of changes

This PR packages [Blockstream/electrs](https://github.com/Blockstream/electrs/), an efficient BTC chain indexer. It forms part of the backend for <https://blockstream.info>, but many others also use it.

`Blockstream/electrs` is actually a fork of [`romanz/electrs`](https://github.com/romanz/electrs), though the two projects diverged several years ago. `romanz/electrs` is also packaged in nixpkgs as [`electrs`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/blockchains/electrs/default.nix)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by phlip9** [comment]
Looks like some of the upstream's dev-dependencies (electrumd, elementsd) fail to compile on anything not- x86_64-linux even when `*_SKIP_DOWNLOAD=1`, so I've disabled tests on all other platforms.

File: pkgs/by-name/bl/blockstream-electrs/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/bl/blockstream-electrs/package.nix:13]:
```suggestion
  version = "0.4.1-unstable-2024-05-29";
```

File: pkgs/by-name/bl/blockstream-electrs/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/bl/blockstream-electrs/package.nix:43]:
```suggestion
  buildInputs = lib.optionals stdenv.hostPlatform.isDarwin [ darwin.apple_sdk.frameworks.Security ];
```

File: pkgs/by-name/bl/blockstream-electrs/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/bl/blockstream-electrs/package.nix:81]:
```suggestion
    description = "Efficient re-implementation of Electrum server in Rust";
```

**Comment by phlip9** [comment]
@FliegendeWurst Thanks for the review! I also took the opportunity to update the package, since it was broken with the current bitcoind version in nixpkgs.

File: pkgs/by-name/bl/blockstream-electrs/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/bl/blockstream-electrs/package.nix:29]: You can use the new `useFetchCargoVendor = true;` to avoid having to copy the Cargo.lock file here.

**Comment by phlip9** [comment]
Rebased on `master` and removed the vendored `Cargo.lock` in favor of `useFetchCargoVendor`

File: pkgs/by-name/bl/blockstream-electrs/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/bl/blockstream-electrs/package.nix:39]: I am not sure these are need anymore.

**Review by FliegendeWurst — APPROVED** [review]

**Comment by phlip9** [comment]
Removed `buildInputs = lib.optionals stdenv.hostPlatform.isDarwin [darwin.apple_sdk.frameworks.Security];`

**Review by FliegendeWurst — APPROVED** [review]

**Comment by phlip9** [comment]
Anything else blocking this from landing?

File: pkgs/by-name/bl/blockstream-electrs/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/bl/blockstream-electrs/package.nix:26]: The PR got merged, so you can drop the patch now and update to a later commit. After that I can merge :)

**Comment by phlip9** [comment]
- Update 0.4.1-unstable-2024-09-30 -> 0.4.1-unstable-2024-11-25
- Remove `bitcoind-v28.patch` since my PR landed
- Gate `preCheck` on `doCheck` so non-x86_64-linux doesn't accidentally break when it can't build the `bitcoind` and `electrum` integration test dependencies.

```diff
-  preCheck = ''
+  preCheck = lib.optionalString doCheck ''
     export BITCOIND_EXE=${bitcoind}/bin/bitcoind
     export ELECTRUMD_EXE=${electrum}/bin/electrum
   '';

```

**Review by FliegendeWurst — APPROVED** [review]

**Pull request merged**
